genai-perf/tests/test_checkpoint.py

```diff
@@ -0,0 +1,71 @@
+# Copyright 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import unittest
+from unittest.mock import patch
+
+from genai_perf.checkpoint.checkpoint import Checkpoint
+from genai_perf.config.input.config_command import ConfigCommand
+from genai_perf.config.run.results import Results
+from tests.test_utils import create_run_config
+
+
+class TestCheckpoint(unittest.TestCase):
+    ###########################################################################
+    # Setup & Teardown
+    ###########################################################################
+    def setUp(self):
+        self._results = Results()
+
+        for i in range(10):
+            run_config_name = "test_run_config_" + str(i)
+            run_config = create_run_config(
+                run_config_name=run_config_name,
+                model_name="test_model",
+                gpu_power=500 + 10 * i,
+                gpu_utilization=50 - i,
+                throughput=300 - 10 * i,
+                latency=100 - 5 * i,
+            )
+            self._results.add_run_config(run_config)
+
+        self._checkpoint = Checkpoint(
+            config=ConfigCommand("test_model"), results=self._results
+        )
+
+    def tearDown(self):
+        patch.stopall()
+
+    ###########################################################################
+    # Checkpoint Tests
+    ###########################################################################
+    def test_checkpoint_methods(self):
+        """
+        Checks to ensure checkpoint methods work as intended
+        """
+
+        # First ensure that there is no state when a checkpoint doesn't exist
+        self.assertEqual({}, self._checkpoint._state)
+
+        # Then write and read back the Results
+        self._checkpoint.write_to_checkpoint()
+        self._checkpoint._read_from_checkpoint()
+        os.remove(self._checkpoint._create_checkpoint_filename())
```

Comment:
I could, but there is utility in writing this out to a file right now (as I'll need to pass this to the visualize team so they can use this as the basis for there work).

Can you point me to a test that mocks this? I couldn't figure out how to make this work from my google searching...

Reply:
Sure! It depends what calls the functions being tested are executing that we want to replace with a mock call.

There are directory creation mocking calls [here](https://github.com/triton-inference-server/perf_analyzer/blob/a55259530eb6a7e0a3cece768826cb095e33ecb9/genai-perf/tests/test_artifacts.py#L34-L36) and some open/is_file/is_dir calls mocked [here](https://github.com/triton-inference-server/perf_analyzer/blob/a55259530eb6a7e0a3cece768826cb095e33ecb9/genai-perf/tests/test_cli.py#L841-L843). Also slightly different opens [here](https://github.com/triton-inference-server/perf_analyzer/blob/a55259530eb6a7e0a3cece768826cb095e33ecb9/genai-perf/tests/test_json_exporter.py#L268-L278) and [here](https://github.com/triton-inference-server/perf_analyzer/blob/a55259530eb6a7e0a3cece768826cb095e33ecb9/genai-perf/tests/test_file_input_retriever.py#L64-L68).

Reply:
It should be done the same. You're mocking the underlying I/O functions which we don't need to test. So if the write to a file is mocked and written to a variable instead, then you can check that variable to assert it has the expected value. Happy to work through this on a call sometime. It might take a bit more time for the first one, then should be easy to replicate wherever else we need it.

If you need to get this in by your PTO, you could clump this in with your other I/O ticket. So any unit tests that actually use I/O functions will be mocked before this goes into main, but you could still get this in before your PTO.

Reply:
Okay, I'll get this done as part of the follow on story before this is merged to main. I agree this needs more testing and should not be writing to disk.